Improve myuplink tests to reach full coverage for all modules

astrandb opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Improve test to reach 100% coverage for all modules.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

astrandb committed:
Full coverage for application_credentials.py

astrandb committed:
Improve tests

astrandb committed:
Add test of connection error

joostlek commented on tests/components/myuplink/test_init.py:
```diff
@@ -130,3 +164,57 @@ async def test_migrate_config_entry(
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test that an incorrect OAuth2 scope fails."""
+    with patch(
+        "homeassistant.components.myuplink.OAUTH2_SCOPES",
+        [
+            "wrong_scope",
+        ],
+    ):
+        await setup_integration(hass, mock_config_entry)
+
+    assert mock_config_entry.state is ConfigEntryState.SETUP_ERROR
```
This is about the reauth right? Why don't we initialize with a different token in there?

astrandb replied:
In the early days of this integration it was designed with a read-only scope. When the scope was extended to switch, select, and number entites we had to add the write scope and the natural way to do that was to trigger a reauth flow with a ConfigEntryAuthFailed exception, same exception as for other authentication failures. As I remember the test was designed to check the reauth flow by triggering with a scope error for no particular reason...

joostlek replied:
Yes but why don't we add a config entry with that old token and have it happen naturally instead of patching the scopes?

astrandb replied:
Changed now so we have three test of the reuth flow: Normal, old scope, account mismatch.

astrandb replied:
Simplified test for bad scope as requested.

joostlek commented on tests/components/myuplink/test_application_credentials.py:
```diff
@@ -0,0 +1,16 @@
+"""Tests for application credentials module."""
+
+from homeassistant.components.myuplink.application_credentials import (
+    async_get_description_placeholders,
+)
+from homeassistant.core import HomeAssistant
+
+
+async def test_placeholders(hass: HomeAssistant) -> None:
+    """Test that required placeholders are returned."""
+    placeholders = await async_get_description_placeholders(hass)
+
+    assert len(placeholders) == 3
+    assert "more_info_url" in placeholders
+    assert "create_creds_url" in placeholders
+    assert "callback_url" in placeholders
```
I am wondering why this doesn't get touched by the config flow tests. I'd rather do that than this

astrandb replied:
Well, I could not find a way to trigger async_get_placeholders(), so I chose this simple approach. Have you got any suggestions?

joostlek replied:
I am also wondering, does it work? Do these URLs show up in the frontend? They might be using some websocket thing to fetch these. I am scared that if we implement this test that this will be a loose function

astrandb replied:
Yes, it works:
![image](https://github.com/user-attachments/assets/258899eb-62da-46a4-b785-1a0837454b09)

astrandb replied:
I ran coverage tests for a few other integrations that implement these placeholders and they don't catch one code line. This leads to 86% coverage in this short little file.

joostlek replied:
Let's leave this open for now then and remove this file for the time being, once I have time to start chasing the quality scale for my integrations I will definitely check this

astrandb committed:
Remove test_application_credentials.py for now.

astrandb committed:
Add more tests of reauth flow

astrandb committed:
Move to snapshots for testing sensor states

astrandb committed:
Simplify test for bad OAuth2 scope

astrandb commented:
I think I have addressed all comments. Coverage is 100% except for application_credentials.
Snapshot tests are implemented for sensors and binary sensors. 
Remaining platforms will be updated in an upcoming PR.

joostlek commented on tests/components/myuplink/test_config_flow.py:
```diff
@@ -158,3 +158,161 @@ async def test_flow_reauth(
+
+
+@pytest.mark.usefixtures("current_request_with_host")
+async def test_flow_reauth_account_mismatch(
```
Can we maybe paremetrize this test and `test_flow_reauth_bad_scope` by making the scope a parameter? That would remove duplication, WDYT?

astrandb replied:
OK, collapsing to one reauth flow test instead of three.

joostlek replied:
Keep in mind that we don't want if statements in tests

joostlek replied:
Ah yes this works beautifully

astrandb committed:
Parametrize to reduce code
